substrate-archive/src/archive.rs

```diff
@@ -32,129 +32,147 @@ use sp_runtime::{
+#[async_trait::async_trait(?Send)]
+pub trait Archive<B: BlockT + Unpin, D: ReadOnlyDB>
+where
+	B::Hash: Unpin,
+{
+	/// start driving the execution of the archive
+	fn drive(&mut self) -> Result<()>;
+
+	/// this method will block indefinitely
+	async fn block_until_stopped(&self) -> ();
+
+	/// shutdown the system
+	fn shutdown(self) -> Result<()>;
+
+	/// Shutdown the system when self is boxed (useful when erasing the types of the runtime)
+	fn boxed_shutdown(self: Box<Self>) -> Result<()>;
+
+	/// Get a reference to the context the actors are using
+	fn context(&self) -> Result<ActorContext<B, D>>;
+}
+
+pub struct ArchiveBuilder<B, R, D, DB> {
+	_marker: PhantomData<(B, R, D, DB)>,
+	/// Chain spec describing the chain
+	chain_spec: Option<Box<dyn ChainSpec>>,
 	/// Path to the rocksdb database
-	pub chain_data_path: Option<String>,
+	chain_data_path: Option<PathBuf>,
 	/// url to the Postgres Database
-	pub pg_url: Option<String>,
+	pg_url: Option<String>,
 	/// how much Cache should Rocksdb Keep
-	pub cache_size: Option<usize>,
+	cache_size: usize,
 	/// number of threads to spawn for block execution
-	pub block_workers: Option<usize>,
+	block_workers: usize,
 	/// Number of 64KB Heap pages to allocate for wasm execution
-	pub wasm_pages: Option<u64>,
-	/// Chain spec describing the chain
-	pub chain_spec: Option<Box<dyn ChainSpec>>,
-	pub _marker: PhantomData<(B, R, D, DB)>,
+	wasm_pages: u64,
 	/// maximum amount of blocks to index at once
-	pub max_block_load: Option<u32>,
+	max_block_load: u32,
 }
 
-impl<B, R, D, DB> Default for Builder<B, R, D, DB> {
+impl<B, R, D, DB> Default for ArchiveBuilder<B, R, D, DB> {
 	fn default() -> Self {
+		let num_cpus = num_cpus::get();
 		Self {
+			_marker: PhantomData,
+			chain_spec: None,
 			chain_data_path: None,
-			cache_size: None,
 			pg_url: None,
-			block_workers: None,
-			wasm_pages: None,
-			chain_spec: None,
-			_marker: PhantomData,
-			max_block_load: None,
+			cache_size: 128,                  // 128 MB
+			block_workers: num_cpus,          // the number of logical cpus in the system
+			wasm_pages: 64 * num_cpus as u64, // 64 * (number of logic cpu's)
+			max_block_load: 100_000,          // 100_000 blocks to index at once
 		}
 	}
 }
 
-impl<B, R, D, DB> Builder<B, R, D, DB> {
+impl<B, R, D, DB> ArchiveBuilder<B, R, D, DB> {
+	/// Specify a chain spec for storing metadata about the running archiver
+	/// in a persistant directory.
+	///
+	/// # Default
+	/// Defaults to storing metadata in a temporary directory.
+	pub fn chain_spec(mut self, spec: Box<dyn ChainSpec>) -> Self {
+		self.chain_spec = Some(spec);
+		self
+	}
+
 	/// Set the chain data backend path to use for this instance.
 	///
 	/// # Default
 	/// defaults to the environment variable CHAIN_DATA_DB
-	pub fn chain_data_db<S: Into<String>>(mut self, path: S) -> Self {
-		self.chain_data_path = Some(path.into());
+	pub fn chain_data_path<S: Into<PathBuf>>(mut self, path: Option<S>) -> Self {
```

Comment:
I don't like this API i.e., changing it to an `Option` the caller should check that before calling these methods IMO.

Can you explain why this is useful?

Reply:
A big reason is that these parameters either have default value, or can be configured through env var, so that these parameters can not be written in the configuration file

Reply:
To be honest, I don’t like these APIs either. I plan to optimize the file configuration in future PRs.

Reply:
Ok, can you please add an issue for this after merging? Fine to leave it out this PR.